packages/frontend/src/components/common/PivotTable/index.tsx

```diff
@@ -53,63 +83,138 @@ const PivotTable: FC<PivotTableProps> = ({
     getFieldLabel,
     getField,
     className,
+    showSubtotals = false,
     ...tableProps
 }) => {
     const containerRef = useRef<HTMLDivElement>(null);
+    const [grouping, setGrouping] = React.useState<GroupingState>([]);
 
-    const getItemFromAxis = useCallback(
-        (rowIndex: number, colIndex: number) => {
-            const value = data.pivotConfig.metricsAsRows
-                ? last(data.indexValues[rowIndex])
-                : last(data.headerValues)?.[colIndex];
-
-            if (!value || !value.fieldId) throw new Error('Invalid pivot data');
-
-            return getField(value.fieldId);
-        },
-        [
-            data.pivotConfig.metricsAsRows,
-            data.headerValues,
-            data.indexValues,
-            getField,
-        ],
-    );
-
-    const getRowTotalValueFromAxis = useCallback(
-        (total: unknown, colIndex: number): ResultValue => {
-            const value = last(data.rowTotalFields)?.[colIndex];
-
-            if (!value || !value.fieldId) throw new Error('Invalid pivot data');
-            const item = getField(value.fieldId);
+    const hasColumnTotals = data.pivotConfig.columnTotals;
 
-            const formattedValue = formatItemValue(item, total);
+    const hasRowTotals = data.pivotConfig.rowTotals;
 
-            return {
-                raw: total,
-                formatted: formattedValue,
-            };
+    const { columns, columnOrder } = useMemo(() => {
+        let headerInfoForColumns = data.headerValues[0].map(() => ({}));
+        headerInfoForColumns = [
+            ...Array(data.indexValueTypes.length),
+            ...headerInfoForColumns,
+        ];
+
+        const newColumnOrder: string[] = [];
+        if (!hideRowNumbers) newColumnOrder.push(ROW_NUMBER_COLUMN_ID);
+
+        let newColumns = data.retrofitData.firstRowOnly.map((col, colIndex) => {
+            newColumnOrder.push(col.fieldId);
+
+            const item = getField(col.underlyingId || col.baseId);
```

Comment:
@agha4to can `firstRowOnly` be typed in `pivot.ts`? I see it as `any[]` there.

This might cause issues when trying to access its properties - e.g. in this line `getField` needs a `string` argument but we might be trying to use a `column` object that doesn't have either `underlyingId` or `baseId` - seeing `packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts:262`

Reply:
Good suggestion. I got rid of the any-s and simplified things in the process. Please let me know of any other recommendations. Thank you 👍